src/rollup/types.d.ts

```diff
@@ -466,37 +457,38 @@ export type SequentialPluginHooks =
 	| 'renderChunk'
 	| 'transform';
 
-export type ParallelPluginHooks =
-	| 'banner'
-	| 'buildEnd'
-	| 'buildStart'
-	| 'footer'
-	| 'intro'
-	| 'moduleParsed'
-	| 'outro'
-	| 'renderError'
-	| 'renderStart'
-	| 'writeBundle'
-	| 'closeBundle'
-	| 'closeWatcher'
-	| 'watchChange';
+export type ParallelPluginHooks = Exclude<
+	keyof FunctionPluginHooks | AddonHooks,
+	FirstPluginHooks | SequentialPluginHooks
+>;
 
-interface OutputPluginValueHooks {
-	banner: AddonHook;
-	cacheKey: string;
-	footer: AddonHook;
-	intro: AddonHook;
-	outro: AddonHook;
-}
+export type AddonHooks = 'banner' | 'footer' | 'intro' | 'outro';
 
-export interface Plugin extends Partial<PluginHooks>, Partial<OutputPluginValueHooks> {
-	// for inter-plugin communication
-	api?: any;
+type MakeAsync<T extends (...a: any) => any> = (
+	...a: Parameters<T>
+) => ReturnType<T> | Promise<ReturnType<T>>;
+
+type ObjectHook<T, O = Record<string, never>> =
+	| T
+	| ({ handler: T; order?: 'pre' | 'post' | null } & O);
```

Comment:
```suggestion
type ObjectHook<T, O = {}> =
	| T
	| ({ handler: T; order?: 'pre' | 'post' | null } & O);
```

It was breaking the auto type inferring

Reply:
> It was breaking the auto type inferring

What do you mean by that? The reason I used `never` over `unknown` was that otherwise, there was no longer a warning about using properties that did not exist e.g. due to a typo.

Reply:
I am not able to provide a playground link as the type is quite complex, but you can try locally by creating this file:

<img width="690" alt="image" src="https://user-images.githubusercontent.com/11247099/184156220-f0caf486-d73a-45c3-bdf2-00e0382b75d1.png">

The `Record<string, never>` have overridden the types for `hanlder` and `order`.

Other than that, I think having it more loosely typed would allow Vite or other extensions to provide more capability in the future.

Reply:
I see. Maybe we need to go for `{}` even though it is a discouraged type. Will have a look tomorrow.

Reply:
I went for `{}` in the end. That way, you actually get an error "can only specify known properties" when using unknown properties while still permitting extension. Unless you use the `PluginImpl` type. Return types are apparently typed more loosely.